src/gui/codeeditors/qgscodeeditor.cpp

```diff
@@ -1173,3 +1178,239 @@ int QgsCodeInterpreter::exec( const QString &command )
   mState = execCommandImpl( command );
   return mState;
 }
+
+
+int QgsCodeEditor::wrapPosition( int line )
+{
+  // If wrapping is disabled, return -1
+  if ( wrapMode() == WrapNone )
+  {
+    return -1;
+  }
+  // Get the current line
+  if ( line == -1 )
+  {
+    int _index;
+    lineIndexFromPosition( linearPosition(), &line, &_index );
+  }
+
+  // If line isn't wrapped, return -1
+  if ( SendScintilla( SCI_WRAPCOUNT, line ) <= 1 )
+  {
+    return -1;
+  }
+
+  // Get the linear position at the end of the current line
+  const long endLine = SendScintilla( SCI_GETLINEENDPOSITION, line );
+  // Get the y coordinates of the start of the last wrapped line
+  const long y = SendScintilla( SCI_POINTYFROMPOSITION, 0, endLine );
+  // Return the linear position of the start of the last wrapped line
+  return static_cast<int>( SendScintilla( SCI_POSITIONFROMPOINT, 0, y ) );
+}
+
+
+
+void QgsCodeEditor::callTip()
+{
+  if ( callTipsStyle() == CallTipsNone || lexer() == nullptr )
+  {
+    return;
+  }
+
+  QsciAbstractAPIs *apis = lexer()->apis();
+
+  if ( !apis )
+    return;
+
+  int pos, commas = 0;
+  bool found = false;
+  char ch;
+
+  pos = linearPosition();
+
+  // Move backwards through the line looking for the start of the current
+  // call tip and working out which argument it is.
+  while ( ( ch = getCharacter( pos ) ) != '\0' )
+  {
+    if ( ch == ',' )
+      ++commas;
+    else if ( ch == ')' )
+    {
+      int depth = 1;
+
+      // Ignore everything back to the start of the corresponding
+      // parenthesis.
+      while ( ( ch = getCharacter( pos ) ) != '\0' )
+      {
+        if ( ch == ')' )
+          ++depth;
+        else if ( ch == '(' && --depth == 0 )
+          break;
+      }
+    }
+    else if ( ch == '(' )
+    {
+      found = true;
+      break;
+    }
+  }
+
+  // Cancel any existing call tip.
+  SendScintilla( SCI_CALLTIPCANCEL );
+
+  // Done if there is no new call tip to set.
+  if ( !found )
+    return;
+
+  QStringList context = apiContext( pos, pos, ctPos );
+
+  if ( context.isEmpty() )
+    return;
+
+  // The last word is complete, not partial.
+  context << QString();
+
+  ct_cursor = 0;
+  ct_shifts.clear();
+  ct_entries = apis->callTips( context, commas, callTipsStyle(), ct_shifts );
+
+  int nr_entries = ct_entries.count();
+
+  if ( nr_entries == 0 )
+    return;
+
+  if ( callTipsVisible() > 0 && callTipsVisible() < nr_entries )
+  {
+    ct_entries = ct_entries.mid( 0, callTipsVisible() );
+    nr_entries = callTipsVisible();
+  }
+
+  int shift;
+  QString ct;
+
+  int nr_shifts = ct_shifts.count();
+
+  if ( callTipsVisible() < 0 && nr_entries > 1 )
```

Comment:
Is this logic adapted from somewhere?

Reply:
Yes, it is directly adapted from [qsciscintilla.cpp](https://github.com/josephwilk/qscintilla/blob/ad08f27aefb0f8194803cf6bff91e208c21f5c84/Qt4Qt5/qsciscintilla.cpp#L360-L367).